server/asset/eth/eth.go

```diff
@@ -136,12 +145,14 @@ func (eth *Backend) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 		eth.log.Errorf("error adding new best block to cache: %v", err)
 	}
 
+	running := make(chan struct{})
```

Comment:
I thought about that, but `run(ctx)` seemed more correct. Will change however.

Reply:
The primary purpose of `Context` provided by the consumer to `Connect` is to shutdown the `Backend` (in `run`, trigger `eth.shutdown()` and break the block polling loop).  So I agree it's right to have `run(ctx)` syntax for *that purpose*...

However, the `Context` we put in `eth.ctx` is to cancel RPCs.  Even if each method had a Context input, we'd still want a _connection-scoped_ context to cancel the RPCs (unless `eth.shutdown` is good enough for that?).

So although it's awkward to both have a `eth.rpcCtx` and do `eth.run(ctx)`, it seems appropriate unless we know `eth.shutdown` is going to have the same effect in which case we could just give the RPCs the background context.

Perhaps it would be more palatable if the `Backend` constructor created a `Context` and `CancelFunc` just for the RPCs, which is then called when the user cancels the context that was given to `Connect`.

i.e.

```diff
diff --git a/server/asset/eth/eth.go b/server/asset/eth/eth.go
index 917f9415..8604ecdc 100644
--- a/server/asset/eth/eth.go
+++ b/server/asset/eth/eth.go
@@ -81,7 +81,11 @@ type ethFetcher interface {
 // data for quick lookups. Backend implements asset.Backend, so provides
 // exported methods for DEX-related blockchain info.
 type Backend struct {
-       ctx  context.Context
+       // A connection-scoped Context is used to used to cancel active RPCs on
+       // subsystem shutdown.
+       rpcCtx     context.Context
+       cancelRPCs context.CancelFunc
+
        cfg  *config
        node ethFetcher
        // The backend provides block notification channels through the BlockChannel
@@ -102,7 +106,10 @@ var _ asset.Backend = (*Backend)(nil)
 // unconnectedETH returns a Backend without a node. The node should be set
 // before use.
 func unconnectedETH(logger dex.Logger, cfg *config) *Backend {
+       ctx, cancel := context.WithCancel(context.Background())
        return &Backend{
+               rpcCtx:     ctx,
+               cancelRPCs: cancel,
                cfg:        cfg,
                blockCache: newBlockCache(logger),
                log:        logger,
@@ -148,14 +155,12 @@ func (eth *Backend) Connect(ctx context.Context) (*sync.WaitGroup, error) {
                eth.log.Errorf("error adding new best block to cache: %v", err)
        }
 
-       running := make(chan struct{})
        var wg sync.WaitGroup
        wg.Add(1)
        go func() {
-               eth.run(ctx, running)
+               eth.run(ctx)
                wg.Done()
        }()
-       <-running
        return &wg, nil
 }
 
@@ -177,7 +182,7 @@ func (eth *Backend) InitTxSizeBase() uint32 {
 
 // FeeRate returns the current optimal fee rate in gwei / gas.
 func (eth *Backend) FeeRate() (uint64, error) {
-       bigGP, err := eth.node.suggestGasPrice(eth.ctx)
+       bigGP, err := eth.node.suggestGasPrice(eth.rpcCtx)
        if err != nil {
                return 0, err
        }
@@ -214,14 +219,14 @@ func (eth *Backend) Synced() (bool, error) {
        // node.SyncProgress will return nil both before syncing has begun and
        // after it has finished. In order to discern when syncing has begun,
        // check that the best header came in under maxBlockInterval.
-       sp, err := eth.node.syncProgress(eth.ctx)
+       sp, err := eth.node.syncProgress(eth.rpcCtx)
        if err != nil {
                return false, err
        }
        if sp != nil {
                return false, nil
        }
-       bh, err := eth.node.bestHeader(eth.ctx)
+       bh, err := eth.node.bestHeader(eth.rpcCtx)
        if err != nil {
                return false, err
        }
@@ -265,17 +270,13 @@ func (eth *Backend) VerifyUnspentCoin(ctx context.Context, coinID []byte) error
 // run processes the queue and monitors the application context. The supplied
 // running channel will be closed upon setting the context which is used by the
 // rpcclient.
-func (eth *Backend) run(ctx context.Context, running chan struct{}) {
-       eth.ctx = ctx
-
-       // Indicate the backend is ready to accept commands.
-       close(running)
-
+func (eth *Backend) run(ctx context.Context) {
        var wg sync.WaitGroup
        wg.Add(1)
        // Shut down the RPC client on ctx.Done().
        go func() {
                <-ctx.Done()
+               eth.cancelRPCs()
                eth.shutdown()
                wg.Done()
        }()
```

Although I acknowledge this has no practical difference from using the same Context for both run and the RPCs.

And the whole thing is unnecessary if `eth.shutdown` interrupts active RPCs.  I just don't know.

Reply:
included these changes